pkg/controller/fileobserver/observer_linux.go

```diff
@@ -0,0 +1,72 @@
+// +build linux
+
+package fileobserver
+
+import (
+	"github.com/golang/glog"
+	"github.com/sigma/go-inotify"
+)
+
+type linuxObserver struct {
+	watcher   *inotify.Watcher
+	reactors  map[string][]reactorFn
+	isRunning bool
+}
+
+func init() {
+	NewObserver = newLinuxObserver
+}
+
+func newLinuxObserver() (Observer, error) {
+	w, err := inotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	return &linuxObserver{watcher: w}, nil
+}
+
+func (o *linuxObserver) AddReactor(reactor reactorFn, filenames ...string) Observer {
+	for _, f := range filenames {
+		glog.V(5).Infof("Observing file %q", f)
+		o.watcher.AddWatch(f, inotify.IN_CREATE|inotify.IN_DELETE|inotify.IN_MODIFY)
+		o.reactors[f] = append(o.reactors[f], reactor)
+	}
+	return o
+}
+
+func (o *linuxObserver) Run(stopCh <-chan struct{}) {
+	if o.isRunning {
+		panic("an instance of file observer is already running")
+	}
+	o.isRunning = true
+	glog.Info("Starting file observer")
+	go func() {
+		for err := range o.watcher.Error {
+			glog.Errorf("watcher error: %v", err)
+		}
+	}()
+
+	go func() {
+		select {
+		case <-stopCh:
+			glog.Infof("Shutting down file observer")
+			o.watcher.Close()
+			o.isRunning = false
+		}
+	}()
+
+	for event := range o.watcher.Event {
+		reactorFns, hasReactor := o.reactors[event.Name]
+		if !hasReactor {
+			continue
+		}
+		glog.Infof("Detected change to %q", event.Name)
+		for _, reactor := range reactorFns {
+			go func() {
+				if err := reactor(event.Name); err != nil {
+					glog.Infof("Error reacting to %q: %v", err)
+				}
+			}()
```

Comment:
use a wait group. Otherwise you can get nested calls. And maybe even queue events for each reactor.